frontend/src/visualization/shared-ops/fetchDataFromService.tsx

```diff
@@ -0,0 +1,191 @@
+
+const useStyles = makeStyles((theme: Theme) =>
+  createStyles({
+    backdrop: {
+      zIndex: theme.zIndex.drawer + 1,
+      color: '#fff',
+    },
+    backdropContent: {
+      marginTop: theme.spacing(3),
+      display: 'flex',
+      flexDirection: 'column',
+      alignItems: 'center',
+    },
+    backdropCancel: {
+      marginTop: theme.spacing(3),
+    },
+    contentContainer: {
+      padding: theme.spacing(3),
+    },
+  })
+);
+
+type RenderFunction<T> = (t: T, update: () => void) => JSX.Element;
+
+type QueryFunction<
+  TArgs extends [...Array<unknown>, CancellationToken],
+  TResult
+> = (...args: TArgs) => Promise<TResult>;
+
+export interface FetchDataOptions<TArgs extends Array<unknown>, TData> {
+  queryFn: QueryFunction<[...TArgs, CancellationToken], TData>;
+  defaultData?: TData;
+}
+
+/**
+ * Fetches data from service using {@link useAsync} with {@link promiseFn} executeQuery.
+ * If the screen is still loading a {@link JSX.Element}, showing a loading screen, is returned.
+ * if an error occurs a {@link JSX.Element}, showing an error screen, is returned.
+ * If the data to be fetched is undefined a {@link JSX.Element}, showing an error message, is returned.
+ *
+ * @param queryFn - the query the service executes
+ * @param service - the service that executes the query
+ * @param arg - optional, additional argument given to {@link useAsync}
+ * @returns if one of the above cases occured, the corresponding {@link JSX.Element}, otherwise the data to be fetched
+ */
+function fetchDataFromService<TArgs extends Array<unknown>, TData>(
+  queryFn: QueryFunction<[...TArgs, CancellationToken], TData>,
+  content: RenderFunction<TData>,
+  ...args: TArgs
+): JSX.Element;
+
+function fetchDataFromService<TArgs extends Array<unknown>, TData>(
+  options: FetchDataOptions<TArgs, TData>,
+  content: RenderFunction<TData>,
+  ...args: TArgs
+): JSX.Element;
+
+function fetchDataFromService<TArgs extends Array<unknown>, TData>(
+  queryFnOrOptions:
+    | QueryFunction<[...TArgs, CancellationToken], TData>
+    | FetchDataOptions<TArgs, TData>,
+  content: RenderFunction<TData>,
+  ...args: TArgs
+): JSX.Element {
+  const queryFn =
+    typeof queryFnOrOptions === 'function'
+      ? queryFnOrOptions
+      : queryFnOrOptions.queryFn;
+
+  const defaultData =
+    typeof queryFnOrOptions === 'function'
+      ? undefined
+      : queryFnOrOptions.defaultData;
+
+  const classes = useStyles();
+
+  // The component state that contains the cancellation token source used to cancel the query operation.
+  const [loadingCancellationSource] = React.useState(
+    new CancellationTokenSource()
+  );
+
+  const [data, setData] = useState<TData | null>(null);
+  const [error, setError] = useState<Error | null>(null);
+  const [isLoading, setIsLoading] = useState(true);
+  const [needsUpdate, setNeedsUpdate] = useState(false);
+
+  const lastUpdateRef = useRef<number | null>(null);
+
+  useEffect(() => {
+    let mounted = true;
```

Comment:
why do you actually need this boolean?

Reply:
We do not. This is useless and is an artifact of a change that was made. It can be removed.

Reply:
then remove it :D

Reply:
To clarify:
The boolean `mounted` is needed, as it tracks whether the component is still mounted. The variable `lastUpdateRef` is an artifact and is not used anymore.